### PR TITLE
New model manageiq manageiq provider

### PR DESCRIPTION
**Pull request by yaacov**
**SUMMARY**

ManageIQ is an open source management platform for Hybrid IT.

This change is adding:

    manageiq_provider module, responsible for provider management in ManageIQ

All the modules, including docs, tests and usage examples can be found here

Currently, the only requirement for the modules is manageiq-api-client-python

**ISSUE TYPE**

    New Module Pull Request

**COMPONENT NAME**

manageiq_provider.py (module)

**ANSIBLE VERSION**

ansible 2.3.0.0

### CODE REVIEW

**Comment by yaacov** [comment]
@gtanzillo @abellotti (ManageIQ team), @gundalow @thaumos (Ansible team), @zgalor @cben @simon3z (Containers Management team) - please review.

Reviewers from ManageIQ: Gregg Tanzillo (@gtanzillo) and Alberto Bellotti (@abellotti).
Maintainers: @cben @zgalor @yaacov

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (zgalor) [inline lib/ansible/modules/remote_management/manageiq/manageiq_provider.py]: please add an example with setting `role: prometheus`

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (zgalor) [inline lib/ansible/modules/remote_management/manageiq/manageiq_provider.py]: the default metric role for amazon should be `None`

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (zgalor) [inline lib/ansible/modules/remote_management/manageiq/manageiq_provider.py]: `ca_bundle_path` param was replaced by `certificate_authority`. Should be removed

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (zgalor) [inline lib/ansible/modules/remote_management/manageiq/manageiq_provider.py]: /s/group/zone/

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (zgalor) [inline lib/ansible/modules/remote_management/manageiq/manageiq_provider.py]: s/user/provider

**Review by zgalor — CHANGES_REQUESTED** [review]
Nice refactoring of original module. Few comments.

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (yaacov) [reply lib/ansible/modules/remote_management/manageiq/manageiq_provider.py]: 👍 added

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (yaacov) [reply lib/ansible/modules/remote_management/manageiq/manageiq_provider.py]: 👍 changed

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (yaacov) [reply lib/ansible/modules/remote_management/manageiq/manageiq_provider.py]: 👍 right

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (yaacov) [reply lib/ansible/modules/remote_management/manageiq/manageiq_provider.py]: 👍 done

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (yaacov) [reply lib/ansible/modules/remote_management/manageiq/manageiq_provider.py]: 👍 done

**Review by zgalor — APPROVED** [review]
LGTM

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (yaacov) [inline lib/ansible/modules/remote_management/manageiq/manageiq_provider.py]: @cben @simon3z WDYT should be the default metrics_role : 'hawkular' or 'prometheus' ?

we currently default to 'hawkular' but thinking of the future we should default to 'prometheus' ?

**Comment by yaacov** [comment]
@gtanzillo @abellotti please review.

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (yaacov) [reply lib/ansible/modules/remote_management/manageiq/manageiq_provider.py]: changed to default to 'prometheus' if role is not explicitly set.

**Comment by yaacov** [comment]
@cben please review

**Comment by yaacov** [comment]
@abellotti @gtanzillo hi, I need your ack / nack before I start nagging the ansible guys :-)

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:514]: Nice :+1:
If trying to edit a type not supported by this module, `class_name_to_type()` will return None and will not give clear error message?
Similarly if creating new and omitted type, will it give clear error message?

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_provider.py]: naming nit: singular `endpoint` is one of `raw_endpoints` from playbook, plural `endpoints` is in "connection_configurations" format.
can you find less confusing names?  perhaps rename `endpoints` list to `configurations`?
(also consider s/build_endpoints/build_configurations/)

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:423]: :+1: this works around https://github.com/ManageIQ/manageiq/pull/13400 bug, and possibly helps with older API versions that had less fields (as long playbook doesn't specify the new fields).

Subtle: here you omit top-level None fields, in create only inside endpoints.
Does this allow you to do partial edit?

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_provider.py]: shouldn't be required, irrelevant for Amazon providers

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_provider.py]: Both ManageIQ and Amazon have "regions" and "zone".  
IINM, it's zone in manageiq DB and AWS region for Amazon.
=> Please clarify descriptions.
=> Consider `manageiq_zone` and/or `provider_region`?

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_provider.py]: Can you add an Amazon example?

**Review by cben — COMMENTED** [review]
Overall neat :smile_cat: 
This is significant simplification compared to what we had in  https://github.com/dkorn/manageiq-ansible-module/blob/master/library/manageiq_provider.py :-)

1. Doesn't wait for ManageIQ to test provider connection — this was complex & fragile, needs better API https://github.com/ManageIQ/manageiq-api/issues/28
2. Doesn't schedule a refresh, IIUC can only be done after (1) above (and possibly would belong in a separate module).
3. Doesn't attempt to guess whether update is needed, this was incorrect due to https://github.com/ManageIQ/manageiq-api/issues/27
  => Always does edit, forced to always assume `changed: true`.  Can't do better without API improvement.

P.S. multi-endpoint "connection_configurations" API is not fully documented, https://github.com/ManageIQ/manageiq-api/issues/26 is best reference I have.

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (yaacov) [reply lib/ansible/modules/remote_management/manageiq/manageiq_provider.py]: 👍 done, added an Amazon example ( without port and with provider_region)

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (yaacov) [reply lib/ansible/modules/remote_management/manageiq/manageiq_provider.py]: 👍 done, now using the variable "connection_configurations" instead of endpoints.

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (yaacov) [reply lib/ansible/modules/remote_management/manageiq/manageiq_provider.py]: 👍 done :-)

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (yaacov) [reply lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:514]: > will it give clear error message?

yes, we will send it to the API, but the error returned is clear and to the point.

Comment (yaacov) [reply lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:514]: > If trying to edit a type not supported by this module, class_name_to_type() will return None and will not give clear error message?

provider['type'] can only be `supported_providers().keys()` so it should not be `None`, but even if it is ^^ :-)

Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_provider.py]: `supported_providers()` has fallthrough `return None` if it doesn't find (unless I'm reading it wrong)

Comment (cben) [reply lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:514]: `class_name_to_type()` has fallthrough return None if it doesn't find (unless I'm reading it wrong)

**Review by cben — APPROVED** [review]
:+1:

@gtanzillo @abellotti Please review.  We especially need your perspective on whether the module interface would work in the future for other provider types.

**Comment by yaacov** [comment]
@martinpovolny

**Review by gtanzillo — APPROVED** [review]
👍 LGTM

**Comment by yaacov** [comment]
shipit

**Comment by yaacov** [comment]
cc @thaumos @mattclay @bcoca

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (syncrou) [inline lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:182]: @yaacov Can you add and example including the token?

Comment (yaacov) [reply lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:182]: happily :-) added at:
https://github.com/ansible/ansible/pull/28273/files#diff-d282374f7f4e8ce8c63becb85a3130b4R226

**Comment by yaacov** [comment]
shipit

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (syncrou) [reply lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:182]: Awesome Thanks!  👍

**Comment by yaacov** [comment]
@thaumos @mattclay @bcoca please review

File: lib/ansible/modules/remote_management/manageiq/manageiq_provider.py
Comment (yaacov) [reply lib/ansible/modules/remote_management/manageiq/manageiq_provider.py:514]: p.s. added checks for null or bad provider type, checked ... this can happen if we manually tweak the type in the DB, and then send an edit request without explicit type :-)
[EDIT: no this can not happen, but leaving the test, it may help in some other case ... ]

**Comment by yaacov** [comment]
shipit

**Comment by yaacov** [comment]
shipit

**Comment by yaacov** [comment]
@thaumos @mattclay @bcoca please review

**Comment by zgalor** [comment]
shipit

**Comment by yaacov** [comment]
@thaumos @mattclay @bcoca please review

**Comment by samdoran** [comment]
Is there an existing ManageIQ module this one deprecates? If so, do we want to warn on that module to use the new one?

**Comment by samdoran** [comment]
I would like to get this merged, but the highly nested options on the `endpoints` field are a bit of a challenge. Accurately documenting them is the first challenge. In the current state, they will not render properly on docs.ansible.com. Second, suboptions more than one level deep tend to indicate overloading a parameter with too much information.

Would it be possible to simplify the module arguments so that so many parameters are not specified in `endpoints`, and move them to top level arguments instead?

I will see about getting this merged in its current state, but I would like to follow up with another PR to simplify the parameters if possible.

**Comment by samdoran** [comment]
Merged to `devel` for inclusion in 2.4. Thank you!

**Comment by yaacov** [comment]
> I would like to follow up with another PR to simplify the parameters if possible.

@samdoran thanks for the docs touch-up 👍 

I think we can do:
```
endpoints:
  - role: 'default'
    hostname: 'example.com'
    ....
  - role: 'hawuklar'
    auth_key: 'fdhlfkha lkj'
    ...

```
instead of:
```
endpoints:
  metrics:
    role: 'default'
    hostname: 'example.com'
    ....
  alerts:
    role: 'hawuklar'
    auth_key: 'fdhlfkha lkj'
    ...

```

@cben @enoodle @ilackarms @zgalor @simon3z WDYT ?

**Comment by simon3z** [comment]
> @cben @enoodle @ilackarms @zgalor @simon3z WDYT ?

@yaacov I think you have to sync with @gtanzillo because the risk here is that this could be too much container-oriented, what would be the plan for other providers?

Even if the implementation is not there I think we should plan for being able to later accommodate any kind of provider.

**Comment by yaacov** [comment]
> @yaacov I think you have to sync with @gtanzillo because the risk here is that this could be too much container-oriented, what would be the plan for other providers?

@simon3z 

All providers have a "default" endpoint, so we are covered here. "default" is the only endpoint that is "required=true" , the other endpoints are per-provider.

We also have a set of pre-defined defaults for each supported provider:
https://github.com/ansible/ansible/pull/28273/files#diff-d282374f7f4e8ce8c63becb85a3130b4R239
When we add a new supported provider, we can also improve the defaults value.

The pros and cons I see in using a list vs. using a dict are:
Pros of list:
1. users can set any set of endpoint they want ( e.g. we do not limit them to one metrics endpoint )
2. it may be more human understable, and will have less indentation levels.

Cons of list:
1. users can set any set of endpoint they want ( e.g. we do not limit them to one metrics endpoint )
2. it's harder code-wise to set default values per `provider-type/option` because we have more possible options and they are not pre-ordered inside a dict.

I like the dict option better, @cben suggested that users may prefer to have one level options ( e.g. "metrics_hostname" instead of `{ "metrics": {"hostname": "example.com"} }` ) because it's more human lagauge like.

**Comment by cben** [comment]
Correction: I'm not suggesting `metrics_hostname`, `metrics_port`... that
would be bad.
My proposal is dropping the `endpoints:`, lifting these to top level:

default:
  role: 'default'
  hostname: 'example.com'
  ....
alerts:
  role: 'hawuklar'
  auth_key: 'fdhlfkha lkj'
  ...

perhaps with an optional common suffix/prefix to separate those from other
level-options:


default_endpoint:
  role: 'default'
  hostname: 'example.com'
  ....
alerts_endpoint:
  role: 'hawuklar'
  auth_key: 'fdhlfkha lkj'
  ...




2017-08-30 14:34 GMT+03:00 Yaacov Zamir <notifications@github.com>:

**Comment by yaacov** [comment]
@enoodle @ilackarms as the only people I know of that use this module :-)
the options now are:

a. leave this as it is now:
```
endpoints:
   default:
        role: ....
        hostname: ...
   metrics:
        hostname: ....

```
b. use one less level ( as @cben suggested ) with "endpoint" being inferred: 
 ```
default:
      role: ....
      hostname: ...
 metrics:
      hostname: ....

```

this two options sound good to me.

1. do you have an opinion ?
2. do you know of other users who may have an opinion ?

**Comment by simon3z** [comment]
@yaacov personally I like "a" (*leave this as it is now*), but maybe I am missing something. I'll leave this to 
 the feedback of the direct users (@ilackarms @joelddiaz @cben)

**Comment by yaacov** [comment]
> but maybe I am missing something.

@simon3z , the missing item of knowledge is that ansible **docs** do not know how to **display help about** 3rd level of indentation, the module itself works fine.

I think the module is better with this extra level of indentation, but a user looking at the docs may get confused.

see:
http://docs.ansible.com/ansible/devel/manageiq_provider_module.html

this is why @samdoran  asked that we consider removing one level of indentation.

**Comment by simon3z** [comment]
> @simon3z , the missing item of knowledge is that ansible docs do not know how to display help about 3rd level of indentation, the module itself works fine.

@yaacov ah yeah indeed I don't think we should design the API trying to please the docs limitations (unless they were there for a reason, e.g. more than two levels are confusing, etc.)

**Comment by yaacov** [comment]
@samdoran hi,
we can fix the extra `suboptions` ( see above discussion ) but we think that "manageiq-wise" it's better the way it is now, how bad is this issue "ansible-wise" ?

**Comment by samdoran** [comment]
The docs rendering issue is minor — it's more about the module UI and how it behaves compared to other modules.

This is the first module we have that has _three_ layers of suboptions. Will this be a burden on the user, having to specify a complex nested dictionary? Is there any way the options could be simplified/flattened to top level parameters, possibly defining dependency relationships in order to provide helpful warnings if certain combinations are submitted/omitted?

I'm not saying it _has_ to be this way, but we generally try to streamline the module UI so they are simple to use.

This is a community module, and it may be that the community using this module likes this UI. Ultimately it is up to the community — we just try to give guidance so modules have a consistent look and feel.

**Comment by yaacov** [comment]
@samdoran thanks,

@ilackarms @joelddiaz @enoodle @cben , please read @samdoran comment above, please vote for one of this options:

Your feedback is important. 

A.

```
name: ...
type: ...
manageiq_connection:
     url: ...
     token: ...
endpoints:
     default:
         role: ....
         hostname: ...
     metrics:
         hostname: ....
```

B.

```
name: ...
type: ...
manageiq_connection:
     url: ...
     token: ...
default:
     role: ....
     hostname: ...
metrics:
     hostname: ....
```

C.
```
name: ...
type: ...
manageiq_connection:
     url: ...
     token: ...
default_endpoint:
     role: ....
     hostname: ...
metrics_endpoint:
     hostname: ....
```

**Comment by joelddiaz** [comment]
The docs do take a moment to get used to (if you're used to reading Ansible module documents). I can see how the manageiq_connection section can be flattened out. Renaming those params to something like miq_username / miq_password / miq_url / etc makes some sense to me. This is more-or-less how the AWS modules work.

The provider section is tougher to flatten out. The parameter naming would end up encoding the parentage of the params. Something like:

The base provider params:
provider_name
provider_type
provider_hostname
provider_ssl
provider_certificate_authority
provider_auth_key

The metrics:
provider_metrics_hostname
provider_metrics_ssl
provider_metrics_certificate_authority

Alerts:
provider_alerts_hostname
provider_alerts_ssl
provider_alerts_certificate_authority

As far as @yaacov 's A/B/C options, I would go with B. But may I suggest one minor change. Instead of default->role, default->hostname, etc, can it be something besides 'default':

provider:    (<--- or 'provider_endpoint')
  role: ...
  hostname: ...

**Comment by cben** [comment]
I vote B.

As for flattenning `manageiq_connection: foo: ...` to `miq_foo: ...`, that's what we had in https://github.com/dkorn/manageiq-ansible-module/ but IMHO nesting there has 2 benefits:
- many concepts are relevant both to ansible<->MIQ connection and MIQ<->provider (host, port, user, pw, ssl enforcement); user name in manageiq_user was similarly  ambiguos.  IMHO having them under a separate section is a clearer separation than prefixing `miq_foo: ...` (but YMMV).
- same connection settings are likely to be reused accross manageiq module calls.  having all in one dictionary allows building it once and concisely reusing.

**Comment by yaacov** [comment]
@cben Thanks, I also think that's the nicest option :-) , also changed "default" to "provider" as @joelddiaz sugested.

See: #28967

**Pull request merged**
